Return Promise for lazy functions.

PIG208 opened the pull request:
Django uses a proxy method for lazy translation functions that actually returns instances of `Promise` instead of `str`. To be accurate, we should change the signatures for these functions.

## Related issues

<!--
Mark what issues this Pull Request closes or references.

Format is:
- Closes #issue-number
- Refs #issue-number

Example. Refs #0
Documentation: https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/
-->

- Closes #688
- Refs #688

<!--
If you have any feedback, just write it here.

It can be whatever you want!
-->

sobolevn commented on django-stubs/utils/translation/__init__.pyi:
```diff
@@ -3,6 +3,7 @@ from contextlib import ContextDecorator
 from typing import Any, Callable, Optional, Union
 
 from django.core.handlers.wsgi import WSGIRequest
+from django.utils.functional import Promise
```
I think that `Promise` is a new thing, isn't it? Looks like that latest `3.2.x` does not have this primitive: https://github.com/django/django/tree/3.2.6/django/utils

sobolevn replied:
Moreover, even in latest `master` it is just a plain class with no methods or anything: https://github.com/django/django/blob/main/django/utils/functional.py#L68

PIG208 replied:
Hmm.. `Promise` has been there for quite a while: https://github.com/django/django/commit/e4e28d907aee9f0e3977e2c6237406d7746c5719. But you are right that `Promise` is just a plain class. Since the proxy method knows about the `resultclasses`, I guess creating generics like `Promise[str]` would be better?

sobolevn replied:
Maybe we can add these methods to `Promise` as well? https://github.com/django/django/blob/3.2.6/django/utils/functional.py#L84-L191

PIG208 committed:
Add proxied functions for Promise.

Although there is nothing defined in `Promise` itself, the only
instances of `Promise` are created by the `lazy` function, with magic
methods defined on it.

https://github.com/django/django/blob/3.2.6/django/utils/functional.py#L84-L191.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Type the return value of lazy translation functions as Promise.

The return value of the lazy translation functions is a proxied
`Promise` object.
https://github.com/django/django/blob/3.2.6/django/utils/translation/__init__.py#L135-L221.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Mark unicode translation functions for deprecation.

https://docs.djangoproject.com/en/4.0/releases/4.0/#features-removed-in-4-0.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

sobolevn commented on django-stubs/utils/functional.pyi:
```diff
@@ -20,7 +20,25 @@ class cached_property(Generic[_T]):
+    def __prepare_class__(cls) -> None: ...
+    @classmethod
+    def __promise__(cls, method_name: str) -> Callable: ...
+    def __text_cast(self) -> str: ...
```
We 100% do not need private methods.

PIG208 replied:
You are right! These are removed.

sobolevn commented on django-stubs/utils/functional.pyi:
```diff
@@ -20,7 +20,25 @@ class cached_property(Generic[_T]):
+    def __mod__(self, rhs: Any) -> str: ...
+    def __add__(self, other: Any) -> str: ...
+    def __radd__(self, other: Any) -> str: ...
+    def __deepcopy__(self, memo: Any): ...
```
Should return `-> Promise`. Ideally it should return self-type or bound-self-typevar, but `Promise` will be good enough.

PIG208 replied:
Looks like the implementation forces evaluation of the lazy value when any of these methods are called. [`__add__` for example](https://github.com/django/django/blob/main/django/utils/functional.py#L195), returns whatever the casted value's `__add__` returns.

sobolevn commented on django-stubs/utils/functional.pyi:
```diff
@@ -20,7 +20,25 @@ class cached_property(Generic[_T]):
+    def __bytes_cast_encoded(self) -> bytes: ...
+    def __cast(self) -> Union[bytes, str]: ...
+    def __lt__(self, other: Any) -> bool: ...
+    def __hash__(self) -> int: ...
```
I think `__hash__` can be removed.

PIG208 replied:
Removed.

PIG208 committed:
Add _StrPromise as a special type for Promise objects for str.

This allows the user to access methods defined on lazy strings while
still letting mypy be aware of that they are not instances of `str`.

The definitions for some of the magic methods are pulled from typeshed. We need
those definitions in the stubs so that `_StrPromise` objects will work properly
with operators, as refining operator types is tricky with the mypy
plugins API.

The rest of the methods will be covered by an attribute hook.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Implement _StrPromise attribute hook.

This implements an attribute hook that provides type information for
methods that are available on `builtins.str` for `_StrPromise` except
the supported operators. This allows us to avoid copying stubs from the
builtins for all supported methods on `str`.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Allow message being a _StrPromise object for RegexValidator.

One intended usage of lazystr is to postpone the translation of the
error message of a validation error. It is possible that we pass a
Promise (specifically _StrPromise) and only evaluate it when a
ValidationError is raised.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

sobolevn commented on mypy_django_plugin/transformers/functional.py:
```diff
@@ -0,0 +1,40 @@
+
+    str_info = helpers.lookup_fully_qualified_typeinfo(helpers.get_typechecker_api(ctx), f"builtins.str")
+    assert str_info is not None
+    method = str_info.get(method_name)
```
I think that you should use `analyze_member_access` here from `mypy/checkmember.py`

PIG208 replied:
Updated.

PIG208 committed:
Refactor _StrPromise attribtue hook with analyze_member_access.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

sobolevn reviewed (approved):
Let's try it out! Thank you!
